<a href="https://colab.research.google.com/github/hpuma/CSC44800/blob/main/Assignment3/Henry_Puma_Linear_Regression_Assignment3.ipynb#scrollTo=sdAQfDqX_6Lv" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

\(\^Be sure to update this button to point to your notebook instead of the sample notebook\)

In [105]:
# Imports section
import numpy as np
import pandas as pd
from sklearn.pipeline import Pipeline
from sklearn import datasets, linear_model
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split,cross_val_score

## Part 1. Loading the dataset

In [80]:
# 1 Using pandas load the dataset (load remotely, not locally)
# 2.Output the first 15 rows of the data
csv_url = "https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv"
data = pd.read_csv(csv_url)
data.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [81]:
# 3.Display a summary of the table information (number of datapoints, etc.)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## Part 2. Splitting the dataset

In [82]:
# Take the pandas dataset and split it into our features (X) and label (y)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
'''
Independent variable - factors that will contribute to the output of the dependent variable (y)
Here we include ALL the rows, but only take in the Temperature and Mols of KCL columns (index 0:2) <-> index 1, and 2 
''' 
X = data.iloc[:,0:2]
'''
Dependent variable - value that was measured as the output from the independent variables
Like before, we get all the rows from the data but ONLY take in the Size nm^3 column (index 3)
''' 
y = data.iloc[:,2:3]
X.head(), y.head()

(   Temperature °C  Mols KCL
 0             469       647
 1             403       694
 2             302       975
 3             779       916
 4             901        18,
       Size nm^3
 0  6.244743e+05
 1  5.779610e+05
 2  6.196847e+05
 3  1.460449e+06
 4  4.325726e+04)

In [83]:
'''
Split data where we have 90% training and 10% testing
X: shape = (1000, 2)
y: shape = (1000, 1)
'''
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((900, 2), (100, 2), (900, 1), (100, 1))

## Part 3. Perform a Linear Regression

In [84]:
# Use sklearn to train a model on the training set
lr_model = linear_model.LinearRegression().fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
data_point = X_test.iloc[:1].values
prediction = lr_model.predict(data_point)

r2_score = lr_model.score(X_train, y_train)
print("R^2=", r2_score)

R^2= 0.8555399490767951


In [85]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Score explanation
    The lr_model.score() method returns the coefficient of determination from our prediction. In other words, this is the R^2 (Residual squared) value. The R^2 value is an indicator of how good of a job our model does in predicting our data point (in our case, the **data_point value**). 
    Using my linear regression model, I was able to get the value of R^2 to be 0.8606240251784951. An R^2 value has a range of (-1.0 <= R^2 <= 1.0). A score of 1.0 means that our model is perfectly fitted to our data point. Since our R^2 ~ 0.86 and relatively close to 1.0, we can say that my model does an adequate at predicting our data point.

In [86]:
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

In [87]:
prediction2 = lr_model.predict(X_test)
r2_score2 = lr_model.score(X_train, y_train)
coefficients, intercept = lr_model.coef_, lr_model.intercept_
print(f"R^2 = {r2_score2}")
print(f"Coefficents {coefficients}")
print(f"Intercept {intercept}")

R^2 = 0.8555399490767951
Coefficents [[ 873.77642614 1025.05000166]]
Intercept [-415658.25994301]


# H(X) function 

Using the extracted values from above, I was able to formulate the following equation:
$$ (x) = 869.91_{x1} + 1023.50_{x2} - 409461.17$$

## Part 4. Use Cross Validation

In [88]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(lr_model, X, y, cv = 5)
print(f"Mean: {scores.mean()}\nStd dev {scores.std()}")
# Report on their finding and their significance

Mean: 0.8568167899144437
Std dev 0.013466307372096023


# Cross Validation Findings
- With our result above, I was able to find the lowest score to be 0.83918826 and the highest score 0.87051239. The scores remained lightly consistent ( greater than 0.80).
- I was also able to find the Mean of all the scores to be ~ 0.86 with a standard deviation of ~ 0.013.

## Part 5. Using Polynomial Regression

In [116]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
'''
Transforming inpu
'''
poly_transform = PolynomialFeatures(degree=2)

X2 = data.iloc[:,0:2]
print(X2)

new_X = poly_transform.fit_transform(X)

X_train2, X_test2, y_train2, y_test2 = train_test_split(new_X, y, test_size=0.10)


model_pipeline = Pipeline([('poly', PolynomialFeatures(degree=2)),('linear', LinearRegression())])

lr_model2 = model_pipeline.fit(X_train2, y_train2)

prediction3 = lr_model2.predict(X_test2)
r2_score3 = lr_model2.score(X_train2, y_train2)

lin_steps = lr_model2.named_steps['linear']

coefficients2, intercept2 = lin_steps.coef_ , lin_steps.intercept_
# Report on the metrics and output the resultant equation as you did in Part 3.
print(f"R^2 = {r2_score3}")
print(f"Coefficents: {coefficients2}", end="\n\n")
print(f"Intercept: {intercept2}")

     Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]
R^2 = 1.0
Coefficents: [[-2.67367470e-07  2.51592989e-07  6.00000001e+00 -6.94148522e-07
   2.98470045e-08  6.66666686e-01  9.52381118e-03 -2.07536668e-10
   6.00000005e+00 -6.43604053e-07 -1.57504285e-08  6.66666657e-01
   9.52381031e-03 -1.57505141e-08  6.66666657e-01  1.27428223e-12
  -5.80668145e-10 -4.49881711e-10  9.52381031e-03  5.81216776e-10
   4.47644315e-10 -3.72017896e-12  9.53620609e-16 -4.14503831e-16
   3.38543092e-07 -3.38543092e-07  1.94583500e-15  1.70072964e-15]]

Intercept: [8.78647552e-05]


# H(X) function 

Using the extracted values from above, I was able to formulate the following equation:
$$ h(x) = 9.78647552 * 10^{-5} -2.67*10^{-7}x_{1} +  2.52*10^{-7}x_{2} + 6x_{1}^{2} ... $$